### ISSUE TITLE
Unexpected "JsonCsrf attack prevention" on a public GET -- a question

### ISSUE THREAD

**Issue by costa** [sinatra/sinatra#2071]
Off-topic (or, perhaps, on-topic actually): your http://sinatrarb.com/rack-protection/ isn't enforcing SSL and contains broken links.

Now, can a maintainer currently responsible for this "racket protection" please answer me this:

If my app -- or a particular part thereof -- does _NOT_ maintain a user session (with a cookie or any other implicit means -- not in a URL), i.e. URLs are "public", clients are "stateless", then, what is the purpose of this "JsonCsrf attack" protection; that is, what kind of attack are we talking about here? (I'm not even restricting this question to just GET's)

Thank you.

**Comment by dentarg** [comment]
You can report website issues at https://github.com/sinatra/sinatra.github.com/issues

**Comment by dentarg** [comment]
Re: JsonCsrf, I encourage that you do your own research. Many things in Sinatra and rack-protection are very old and it is not certain they make sense today.

**Comment by costa** [comment]
@dentarg Hi Patrik,
Thank you for your encouragement, I've been doing my research -- and development -- for many years now, and that's why I was pretty confident when reporting this issue.

I'm sorry for being unclear with the description above, this was a bug report, that is, a report of incorrect behaviour. Now, I'm not sure how this is relevant if it's an old bug or a newer one, this behaviour could not make sense in any time frame. CSRF protection only makes sense with a session-management middleware -- when there's some kind of session management active.

So, again, please forgive my smart description of the issue, but I don't really think this issue can be closed without as little as mentioning it in the docs -- I've been using Sinatra for more than a decade in dozens of projects, and if this "rack-protection" functionality comes as a surprise to me, it's definitely worth mentioning somewhere at least, if not making properly separate and optional. Let's keep the spirit of the original Sinatra project and keep it as accessible and friendly (and surprise-free) as intended.

**Comment by jkowens** [comment]
@costa not sure if you saw the comments in the code, but it describes the reasoning at the time for this protection:

https://github.com/sinatra/sinatra/blob/7b50a1bbb5324838908dfaa00ec53ad322673a29/rack-protection/lib/rack/protection/json_csrf.rb#L13

http://haacked.com/archive/2008/11/20/anatomy-of-a-subtle-json-vulnerability.aspx

**Comment by costa** [comment]
@jkowens No, I didn't see these comments, however, they actually confirmed my point: if you read my description carefully, I'm talking about *a request with no session*, and the attached article says "... Imagine that you’re logged in to a trusted site..." -- which means **session**. CSRF protection should reasonably be implemented at the level of session-managing middleware -- and turned on only when it's active, of course.

**Comment by costa** [comment]
@dentarg Please see my point above; I'm not sure about other types of attacks -- without reviewing them properly, but CSRF protection should be turned on by a session-managing mechanism only.

**Comment by jkowens** [comment]
Yeah, it does look like we should consider disabling this rack protection if sessions aren't enabled. It seems like it should be handled like `remote_token` here:

https://github.com/sinatra/sinatra/blob/7b50a1bbb5324838908dfaa00ec53ad322673a29/rack-protection/lib/rack/protection.rb#L31C8-L32C21
